activerecord/lib/active_record/associations/has_many_through_association.rb

```diff
@@ -53,16 +53,24 @@ def concat_records(records)
         #
         # However, after insert_record has been called, the cache is cleared in
         # order to allow multiple instances of the same record in an association.
-        def build_through_record(record)
-          @through_records[record] ||= begin
-            ensure_mutable
+        def through_records(record)
+          @through_records[record] ||= build_through_record(record)
+        end
 
-            attributes = through_scope_attributes
-            attributes[source_reflection.name] = record
-            attributes[source_reflection.foreign_type] = options[:source_type] if options[:source_type]
+        def build_through_record(record)
+          ensure_mutable
 
-            through_association.build(attributes)
+          if (inverse = source_reflection.inverse_of) && inverse.has_one? &&
+             (has_one_target = record.association(inverse.name).scope.where(through_scope_attributes).first)
+            has_one_target[through_reflection.foreign_key] = owner.id
```

Comment:
Do we also need to assign `has_one_target[through_reflection.foreign_type]` in some cases?

Reply:
Yeah, probably :disappointed: 
Could you TLDR me what are the policies for changing the models? Could I just make `Post` STI, so `foreign_type` is available and create a new association chain to check case from the previous comment?